### PR TITLE
fix: idioms @valenfv_

### PR DESCRIPTION
**Pull request by valenfv**

### CODE REVIEW

File: profile-specific-challenges/frontend/components/idioms.jsx [hunk 1]
```diff
@@ -52,12 +58,13 @@ export function UseEffect({ fetchURL, label }) {
...
   const [clickedTimes, setClickedTimes] = useState()
 
-  useEffect(() => {
-    setReminder(clickedTimes % 5)
-  }, [clickedTimes])
+  // memoizing this might be an overkill.
+  // the useEffect will trigger the reconciliation algorithm twice, one after setting the clickedItems state
+  // and one after setting the remainder state.
+  // we saved some commits with this derived state
+  const remainder = useMemo(() => clickedTimes % 5, [clickedTimes]);
```
Comment (conanbatt) [inline profile-specific-challenges/frontend/components/idioms.jsx:67]: Should not memo this, its an unnecessary memory hodg. The key is to not cause re-renders for cheap calculations.

File: profile-specific-challenges/frontend/components/idioms.jsx [hunk 2]
```diff
@@ -147,82 +160,99 @@ export function UntraceableState() {
     </div>
   )
 }
-
+// any object has to be outside the component's declaration as it will be recreated on every render
+// if the object is compound of any state or props we would need to go in the memoization path
+const calendarDays = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30];
```
Comment (conanbatt) [inline profile-specific-challenges/frontend/components/idioms.jsx:165]: Incorrect! There are a number of elegant ways to declare this better.

File: profile-specific-challenges/frontend/components/idioms.jsx [hunk 3]
```diff
@@ -232,7 +262,7 @@ export function DangerousIdentifier() {
         <button>Add Person</button>
       </form>
       <ul>
-        {people.map(person => <span key={person.name}>{person.name}</span>)}
+        {people.map(person => <li key={person.name}>{person.name}</li>)}
```
Comment (conanbatt) [inline profile-specific-challenges/frontend/components/idioms.jsx:265]: The key was not fixed!

File: profile-specific-challenges/frontend/components/idioms.jsx [hunk 4]
```diff
@@ -247,8 +277,9 @@ export function UnnecessaryEffectTriggering() {
 
   useEffect(() => {
     const interval = setInterval(async () => {
+      // next line is returning a new object every time,
       const leader = await fetchLeader()
-      setLeader(leader)
+      setLeader(prev => leader.name !== prev.name ? leader : prev)
```
Comment (conanbatt) [inline profile-specific-challenges/frontend/components/idioms.jsx:282]: Does react recognize there is no state changes here? I think this probably triggers a re-render anyway.

File: profile-specific-challenges/frontend/components/idioms.jsx [hunk 5]
```diff
@@ -346,15 +378,19 @@ export function UnoptimizableRenderingStructure(altRecords) {
     loadRecords();
   }, []); 
 
+  // you can either do this or use the React.memo HOC by creating another component
+  const memoizedAltRecords = useMemo(() => (
+    <ul>
+      {altRecords.map(rec => <li key={rec.id}>{rec.id}</li>)}
+    </ul>
+  ), [altRecords]);
```
Comment (conanbatt) [inline profile-specific-challenges/frontend/components/idioms.jsx:386]: Just like at the beginning of this file (components as variables) I don't like jsx named used this way. 
Also this useMemo is also a memory tradeoff for all possible alt records. Probably not good. I think it should just be a different react component.

File: profile-specific-challenges/frontend/components/idioms.jsx (hunk 4 above)
Comment (valenfv) [reply profile-specific-challenges/frontend/components/idioms.jsx:282]: I had the same doubt but seems like the param shares the same memory address.

File: profile-specific-challenges/frontend/components/idioms.jsx (hunk 1 above)
Comment (valenfv) [reply profile-specific-challenges/frontend/components/idioms.jsx:67]: You are right; I overlooked the logic

**Pull request closed without merging**